src/base/compressor.hpp

```diff
@@ -0,0 +1,173 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * compressor.hpp
+ */
+
+#ifndef _BASE_COMPRESSOR_HPP_
+#define _BASE_COMPRESSOR_HPP_
+
+#include "lz4.h"
+#include "error_manager.h"
+
+#include <memory>
+#include <variant>
+
+namespace cubcompress
+{
+  struct LZ4 {};
+  /* add here */
+
+  /* options */
+  struct lz4_options
+  {
+    int accel = 1;
+  };
+  /* add here */
+
+  using options = std::variant<lz4_options /*, add here */>;
+
+  /* ---------------------------------------------------------------------------- */
+  /* LZ4									  */
+  /* ---------------------------------------------------------------------------- */
+  struct LZ4ContextDeleter
+  {
+    void operator () (LZ4_stream_t *ctx) const noexcept
+    {
+      if (ctx)
+	{
+	  (void) LZ4_freeStream (ctx);
+	}
+    }
+  };
+
+  inline LZ4_stream_t *lz4_stream ()
+  {
+    thread_local std::unique_ptr<LZ4_stream_t, LZ4ContextDeleter> context { nullptr };
+
+    if (!context)
+      {
+	context.reset (LZ4_createStream ());
+	if (!context)
+	  {
+	    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 1, "failed to create LZ4 stream context");
+	  }
+      }
+    return context.get ();
+  }
+
+  inline int lz4_bound (int size)
+  {
+    return LZ4_compressBound (size);
+  }
+
+  inline int lz4_compress (const char *src, char *dst, int src_size, int dst_capacity, int accel)
+  {
+    LZ4_stream_t *ctx;
+    int compressed_size;
+
+    ctx = lz4_stream ();
+    if (!ctx)
+      {
+	/* failed to get lz4 stream (context) */
+	return 0;
+      }
+
+    LZ4_resetStream_fast (ctx);
+
+    compressed_size = LZ4_compress_fast_extState (ctx, src, dst, src_size, dst_capacity, accel);
+    if (compressed_size <= 0)
+      {
+	/* failed to compress */
+	er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 1, "failed to compress with LZ4_compress_fast_extState");
+      }
+    return compressed_size;
+  }
+
+  inline int lz4_decompress (const char *src, char *dst, int compressed_size, int dst_capacity)
+  {
+    int decompressed_size;
+
+    decompressed_size = LZ4_decompress_safe (src, dst, compressed_size, dst_capacity);
+    if (decompressed_size < 0)
+      {
+	/* failed to decompress */
+	er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 1, "failed to decompress with LZ4_decompress_safe");
+      }
+    return decompressed_size;
+  }
+
+  /* ---------------------------------------------------------------------------- */
+  /* interface									  */
+  /* ---------------------------------------------------------------------------- */
+  template <typename T>
```

Comment:
디폴트 타입을 LZ4로 지정하는 것도 괜찮을 것 같습니다.
```suggestion
  template <typename T = LZ4>
```

Reply:
LZ4의 typename은 caller에 명시되어 있는 편이 좋아 보일 것 같습니다.